File: GuardianMonsters/inventory/src/de/limbusdev/guardianmonsters/inventory/ItemsSubMenu.java
package de.limbusdev.guardianmonsters.inventory;

import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ArrayMap;

import de.limbusdev.guardianmonsters.guardians.items.Inventory;
import de.limbusdev.guardianmonsters.guardians.items.Item;
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;
import de.limbusdev.guardianmonsters.guardians.monsters.Team;
import de.limbusdev.guardianmonsters.inventory.ui.widgets.items.ItemApplicationWidget;
import de.limbusdev.guardianmonsters.inventory.ui.widgets.items.ItemCategoryToolbar;
import de.limbusdev.guardianmonsters.inventory.ui.widgets.items.KeyItemDetailViewWidget;
import de.limbusdev.guardianmonsters.inventory.ui.widgets.items.WeaponDetailViewWidget;
import de.limbusdev.guardianmonsters.ui.widgets.ItemListWidget;

/**
 * Created by Georg Eckert on 17.02.17.
 */

public class ItemsSubMenu extends AInventorySubMenu
    implements ItemCategoryToolbar.ClickListener, ItemListWidget.ClickListener {

    private Inventory inventory;
    private ItemApplicationWidget detailView;
    private Team team;
    private ItemListWidget itemListWidget;

    public ItemsSubMenu(Skin skin, Inventory inventory, Team team) {

        super(skin);
        this.team = team;
        this.inventory = inventory;

        ItemCategoryToolbar toolbar = new ItemCategoryToolbar(skin, this);
        addActor(toolbar);

        // ......................................................................... SCROLLABLE LIST
        Array<Item.Category> filters = new Array<>();
        filters.add(Item.Category.MEDICINE);
        itemListWidget = new ItemListWidget(inventory, this, filters);
        itemListWidget.setPosition(68,0,Align.bottomLeft);

        addActor(itemListWidget);
    }



    private void showItemDetailView(Item item)  {

        if(detailView != null) {

            detailView.remove();
        }

        switch(item.getCategory()) {
            case EQUIPMENT:
                detailView = new WeaponDetailViewWidget(inventory, team);
                break;
            case KEY:
                detailView = new KeyItemDetailViewWidget(inventory, team);
                break;
            default:
                detailView = new ItemApplicationWidget(inventory, team);
                break;
        }
        detailView.setPosition(264,2, Align.bottomLeft);
        detailView.initialize(item);
        addActor(detailView);
    }

    // .............................................................................. CLICK LISTENER

    @Override
    public void onMedicineButton() {
        itemListWidget.applyFilter(Item.Category.MEDICINE);
    }

    @Override
    public void onOtherItemsButton() {
        itemListWidget.applyFilter(Item.Category.ALL);
    }

    @Override
    public void onEquipItemsButton() {
        itemListWidget.applyFilter(Item.Category.EQUIPMENT);
    }

    @Override
    public void onKeyItemsButton() {
        itemListWidget.applyFilter(Item.Category.KEY);
    }

    @Override
    public void onChoosingItem(Item item) {
        showItemDetailView(item);
    }

    @Override
    public void refresh() {
        // Do nothing
    }

    @Override
    protected void layout(Skin skin) {

    }
}

File: GuardianMonsters/inventory/src/de/limbusdev/guardianmonsters/inventory/ItemsSubMenu.kt
package de.limbusdev.guardianmonsters.inventory

import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ArrayMap

import de.limbusdev.guardianmonsters.guardians.items.Inventory
import de.limbusdev.guardianmonsters.guardians.items.Item
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian
import de.limbusdev.guardianmonsters.guardians.monsters.Team
import de.limbusdev.guardianmonsters.inventory.ui.widgets.items.ItemApplicationWidget
import de.limbusdev.guardianmonsters.inventory.ui.widgets.items.ItemCategoryToolbar
import de.limbusdev.guardianmonsters.inventory.ui.widgets.items.KeyItemDetailViewWidget
import de.limbusdev.guardianmonsters.inventory.ui.widgets.items.WeaponDetailViewWidget
import de.limbusdev.guardianmonsters.services.Services
import de.limbusdev.guardianmonsters.ui.widgets.ItemListWidget

/**
 * ItemsSubMenu allows the user to view, choose, apply and organize
 * [Item]s, [Equipment] and [KeyItem]s.
 *
 * @author Georg Eckert 2017-02-17
 */

class ItemsSubMenu
(
        skin: Skin,
        private val inventory: Inventory,
        private val team: Team
)
    : AInventorySubMenu(), ItemCategoryToolbar.ClickListener, ItemListWidget.ClickListener
{
    // --------------------------------------------------------------------------------------------- PROPERTIES
    private var detailView      : ItemApplicationWidget = ItemApplicationWidget(inventory, team)
    private val itemListWidget  : ItemListWidget
    private val toolbar = ItemCategoryToolbar(skin, this)


    // --------------------------------------------------------------------------------------------- CONSTRUCTORS
    init
    {
        // .......................................... scrollable list
        // initialize with medical items
        val filters = Array<Item.Category>()
        filters.add(Item.Category.MEDICINE)

        itemListWidget = ItemListWidget(inventory, this, filters)
        itemListWidget.setPosition(68f, 0f, Align.bottomLeft)

        // Add widgets
        addActor(toolbar)
        addActor(itemListWidget)
    }


    // --------------------------------------------------------------------------------------------- METHODS
    private fun showItemDetailView(item: Item)
    {
        // Remove currently displayed ItemDetailView
        detailView.remove()

        detailView = when (item.category)
        {
            Item.Category.EQUIPMENT -> WeaponDetailViewWidget(inventory, team)
            Item.Category.KEY       -> KeyItemDetailViewWidget(inventory, team)
            else                    -> ItemApplicationWidget(inventory, team)
        }

        detailView.initialize(item)

        detailView.setPosition(264f, 2f, Align.bottomLeft)
        addActor(detailView)
    }

    // .............................................................................. CLICK LISTENER

    override fun onMedicineButton()
    {
        itemListWidget.applyFilter(Item.Category.MEDICINE)
    }

    override fun onOtherItemsButton()
    {
        itemListWidget.applyFilter(Item.Category.ALL)
    }

    override fun onEquipItemsButton()
    {
        itemListWidget.applyFilter(Item.Category.EQUIPMENT)
    }

    override fun onKeyItemsButton()
    {
        itemListWidget.applyFilter(Item.Category.KEY)
    }

    override fun onChoosingItem(item: Item)
    {
        showItemDetailView(item)
    }

    override fun refresh()
    {
        // Do nothing
    }

    override fun layout(skin: Skin) {}
}
